crates/ruff_linter/src/rules/pylint/rules/no_method_decorator.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+/// PLR0203
+pub(crate) fn no_staticmethod_decorator(checker: &mut Checker, class_def: &ast::StmtClassDef) {
+    get_undecorated_methods(checker, class_def, &MethodType::Staticmethod);
+}
+
+fn get_undecorated_methods(
+    checker: &mut Checker,
+    class_def: &ast::StmtClassDef,
+    method_type: &MethodType,
+) {
+    let mut explicit_decorator_calls: HashMap<String, TextRange> = HashMap::default();
+
+    let (method_name, diagnostic_type): (&str, DiagnosticKind) = match method_type {
+        MethodType::Classmethod => ("classmethod", NoClassmethodDecorator.into()),
+        MethodType::Staticmethod => ("staticmethod", NoStaticmethodDecorator.into()),
+    };
+
+    // gather all explicit *method calls
+    for stmt in &class_def.body {
+        if let Stmt::Assign(ast::StmtAssign { targets, value, .. }) = stmt {
+            if let Expr::Call(ast::ExprCall {
+                func, arguments, ..
+            }) = value.as_ref()
+            {
+                if let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() {
+                    if id == method_name && checker.semantic().is_builtin(method_name) {
+                        if arguments.args.len() != 1 {
+                            continue;
+                        }
+
+                        if targets.len() != 1 {
+                            continue;
+                        }
+
+                        let target_name = match targets.first() {
+                            Some(Expr::Name(ast::ExprName { id, .. })) => id.to_string(),
+                            _ => continue,
+                        };
+
+                        if let Expr::Name(ast::ExprName { id, .. }) = &arguments.args[0] {
+                            if target_name == *id {
+                                explicit_decorator_calls.insert(id.clone(), stmt.range());
+                            }
+                        };
+                    }
+                }
+            }
+        };
+    }
+
+    if explicit_decorator_calls.is_empty() {
+        return;
+    };
+
+    for stmt in &class_def.body {
+        if let Stmt::FunctionDef(ast::StmtFunctionDef {
+            name,
+            decorator_list,
+            ..
+        }) = stmt
+        {
+            if !explicit_decorator_calls.contains_key(name.as_str()) {
+                continue;
+            };
+
+            // if we find the decorator we're looking for, skip
+            if decorator_list.iter().any(|decorator| {
+                if let Expr::Name(ast::ExprName { id, .. }) = &decorator.expression {
+                    if id == method_name && checker.semantic().is_builtin(method_name) {
+                        return true;
+                    }
+                }
+
+                false
+            }) {
+                continue;
+            }
+
+            let mut diagnostic = Diagnostic::new(
+                diagnostic_type.clone(),
+                TextRange::new(stmt.range().start(), stmt.range().start()),
```

Comment:
I don't really know what range we should highlight with the diagnostic, please advise!